### PR TITLE
Add support for shadow DOM #604

### PR DESCRIPTION
**Pull request by dgr**
Please complete and include the following checklist:

- [x ] I have read [CONTRIBUTING](https://github.com/clj-commons/etaoin/blob/master/CONTRIBUTING.md) and the [Etaoin Developer Guide](https://github.com/clj-commons/etaoin/blob/master/doc/02-developer-guide.adoc).

- [x ] This PR corresponds to an issue that the Etaoin maintainers have agreed to address. 

- [x ] This PR contains test(s) to protect against future regressions

- [x ] I have updated [CHANGELOG.adoc](https://github.com/clj-commons/etaoin/blob/master/CHANGELOG.adoc) with a description of the addressed issue.

### CODE REVIEW

**Comment by lread** [comment]
Nice work @dgr! Thanks for taking so much care!

Some thoughts/ideas/observations:

Maybe docstrings could link to Etaoin user guide which in turn could link to https://wpt.fyi/results/webdriver/tests/classic/find_element_from_shadow_root/find.py?label=experimental&label=master&aligned in a prominent adoc NOTE. This way users would get a link to your very nice overview of the shadow dom (I read, I learned!) AND a link to current implementation limitations. Whaddya think?

Test-doc is failing because code blocks will return different results each time they are run. For example, this gets converted into a test:
```clojure
(e/get-element-shadow-root driver {:id "shadow-root-host"})
;; => "ac5ab914-7f93-427f-a0bf-f7e91098fd37"
```
which expects a return of `"ac5ab914-7f93-427f-a0bf-f7e91098fd37"` always and will therefore fail.

Options:
1. Add a `//:test-doc-blocks/skip` before the code block to have test-doc-blocks skip the code block
2. Remove the `;; =>` lines from the code example, test-doc-block will only expect the code to run without exception and not expect a specific return. If you feel it important the reader of the doc see a typical return, you could instead express that in a comment or descriptive text outside of the code block
3. In some cases, consider rewriting the code sample, for example this:
    ```clojure
    (e/query-shadow-root driver {:id "shadow-root-host"} {:css "#in-shadow"})
    ;; => "30fca382-6d4a-4f8a-9534-db76a1ed7cba"
    (e/get-element-text-el driver "30fca382-6d4a-4f8a-9534-db76a1ed7cba")
    ;; => "I'm in the shadow DOM"
    ```
    Could become:
    ```clojure
    (->> (e/query-shadow-root driver {:id "shadow-root-host"} {:css "#in-shadow"})
         (e/get-element-text-el driver))
    ;; => "I'm in the shadow DOM"
    ```

I reviewed the user guide for what strategy I took for the element id returns, and it looks like I entirely avoided it in code blocks. It seems [I spent a little bit of time upfront describing the difference between -el and non-el variants](https://github.com/clj-commons/etaoin/blob/master/doc/01-user-guide.adoc#selecting-elements) and didn't dwell on it at all later. 

For me, it is a balance between "what does the reader need to see to understand this?" and "can I ensure that the code blocks should always work in the reader's REPL when they try them out?".

Anyhoo, great work, and I am happy to work out the above with you further.

**Comment by dgr** [comment]
OK, let me do one more commit with updates to doc strings and doc-test stuff.

File: src/etaoin/api.clj [hunk 1]
```diff
@@ -1611,6 +1611,108 @@
...
+  "Returns the shadow root for the specified element or `nil` if the
+  element does not have a shadow root."
+  [driver el]
+  (-> (get-element-property-el driver el "shadowRoot")
+      first
+      second))
+
+(defn get-element-shadow-root
+  "Returns the shadow root for the first element matching the query, or
+  `nil` if the element does not have a shadow root."
```
Comment (lread) [inline src/etaoin/api.clj:1624]: I think in all other fns with `q` I've included:

See [[query]] for details on `q`.

File: src/etaoin/api.clj [hunk 2]
```diff
@@ -1611,6 +1611,108 @@
...
+  For details on `q`, see [[query]].
+
+  The `shadow-q` parameter is similar to the `q` parameter of
+  the [[query]] function, but some drivers may limit it to specific
+  formats (e.g., CSS). See [this note](/doc/01-user-guide.adoc#shadow-root-browser-limitations] for more information."
+  [driver q shadow-q]
+  (query-shadow-root-el driver (get-element-shadow-root driver q) shadow-q))
+
+(defn query-all-shadow-root
+  "First, conducts a standard search (as if by `query`) for an element
```
Comment (lread) [inline src/etaoin/api.clj:1704]: If you change `` `query` `` to `[[query]]`, cljdoc will link to the query fn.

File: src/etaoin/api.clj [hunk 3]
```diff
@@ -1611,6 +1611,108 @@
...
+
+  The `shadow-q` parameter is similar to the `q` parameter of
+  the [[query]] function, but some drivers may limit it to specific
+  formats (e.g., CSS). See [this note](/doc/01-user-guide.adoc#shadow-root-browser-limitations] for more information."
+  [driver shadow-root-el shadow-q]
+  (let [[loc term] (query/expand driver shadow-q)]
+    (find-elements-from-shadow-root* driver shadow-root-el loc term)))
+
+(defn query-shadow-root
+  "First, conducts a standard search (as if by `query`) for an element
```
Comment (lread) [inline src/etaoin/api.clj:1691]: If you change `` `query` `` to `[[query]]`, cljdoc will link to the query fn.

File: src/etaoin/api.clj [hunk 4]
```diff
@@ -1611,6 +1611,108 @@
...
+  with a shadow root. Then, from that shadow root element, conducts a
+  search of the shadow DOM for the first element matching `shadow-q`.
+
+  For details on `q`, see [[query]].
+
+  The `shadow-q` parameter is similar to the `q` parameter of
+  the [[query]] function, but some drivers may limit it to specific
+  formats (e.g., CSS). See [this note](/doc/01-user-guide.adoc#shadow-root-browser-limitations] for more information."
+  [driver q shadow-q]
```
Comment (lread) [inline src/etaoin/api.clj:1700]: Small typo in link syntax, needs to close with `)` instead of `]`

File: src/etaoin/api.clj [hunk 5]
```diff
@@ -1611,6 +1611,108 @@
...
+(defn query-all-shadow-root
+  "First, conducts a standard search (as if by `query`) for an element
+  with a shadow root. Then, from that shadow root element, conducts a
+  search of the shadow DOM for all elements matching `shadow-q`.
+
+  For details on `q`, see [[query]].
+
+  The `shadow-q` parameter is similar to the `q` parameter of
+  the [[query]] function, but some drivers may limit it to specific
+  formats (e.g., CSS). See [this note](/doc/01-user-guide.adoc#shadow-root-browser-limitations] for more information."
```
Comment (lread) [inline src/etaoin/api.clj:1712]: Small typo in link syntax, needs to close with `)` instead of `]`

File: src/etaoin/api.clj
Comment (lread) [inline src/etaoin/api.clj:1]: The namespace docstring includes a toc of sorts. You'll want to add your new fns in there. 
Probably under **Querying/Selecting DOM Elements**

File: doc/01-user-guide.adoc [hunk 6]
```diff
@@ -845,6 +845,125 @@ The following query will find a vector of `div` tags, then return a set of all `
...
+The `query-shadow-root-el` and `query-all-shadow-root-el` allow you to specify the shadow root host element directly, rather than querying for it.
+
+[source,clojure]
+----
+(def in-shadow (e/query-shadow-root driver {:id "shadow-root-host"} {:css "#in-shadow"}))
+(e/get-element-text-el driver in-shadow)
+;; => "I'm in the shadow DOM"
+
+(->> (e/query-all-shadow-root driver {:id "shadow-root-host"} {:css "span"})
+           (map #(e/get-element-text-el driver %)))
```
Comment (lread) [inline doc/01-user-guide.adoc:944]: minor: indentation

File: doc/01-user-guide.adoc [hunk 7]
```diff
@@ -845,6 +845,125 @@ The following query will find a vector of `div` tags, then return a set of all `
...
+;; => ("I'm in the shadow DOM" "I'm also in the shadow DOM")
+
+(def shadow-root (e/get-element-shadow-root-el driver host))
+(e/get-element-text-el driver (e/query-shadow-root-el driver shadow-root {:css "#in-shadow"}))
+;; => "I'm in the shadow DOM"
+ 
+(->> (e/query-all-shadow-root-el driver shadow-root {:css "span"})
+           (map #(e/get-element-text-el driver %)))
```
Comment (lread) [inline doc/01-user-guide.adoc:952]: minor: identation

**Review by lread — CHANGES_REQUESTED** [review]
Getting real close @dgr! Just a few minor things I noticed.

**Comment by dgr** [comment]
OK, try it again.

**Review by lread — APPROVED** [review]

**Comment by lread** [comment]
Looks great @dgr, thanks!

**Pull request merged**